### ISSUE TITLE
[css-selectors-4] Clarify the grammar of a functional pseudo-class

### ISSUE THREAD

**Issue by cdoublev** [w3c/csswg-drafts#7016]
A (list of) selector is parsed with *parse something according to a CSS grammar*:

> Selectors are [parsed](https://drafts.csswg.org/css-syntax-3/#css-parse-something-according-to-a-css-grammar) according to the following grammar:
> - `<selector-list> = <complex-selector-list>`
> - [...]
> - `<pseudo-class-selector> = ':' <ident-token> | ':' <function-token> <any-value> ')'`

With `:not(.class)` as the input, *parse something according to a CSS grammar* normalizes the input then runs *parse a list of component values*, which runs *consume a component value*, which runs *consume a function*, which creates a function component value when reading the `<function-token>` with value `not`, consumes `)`, consumes the following tokens as the (arguments) value of the function component value, and returns the function component value after reading `)`.

`:not(.class)` would be consumed to the component value eg. `{ name: 'not', value: [{ type: 'ident-token', value: 'class' }] }`. Therefore there will be no `<function-token>` when matching `<pseudo-class-selector>` (the production for a functional pseudo-class, I think). 

> The syntax of a pseudo-class consists of a ":" (U+003A COLON) followed by the name of the pseudo-class as a CSS identifier, and, in the case of a functional pseudo-class, a pair of parentheses containing its arguments.

I think `<pseudo-class-selector>` could be defined in prose instead, and its definition could re-use the wording for [`<url-modifier>`](https://drafts.csswg.org/css-values-4/#url-modifiers), which refers to the definition of *a functional notation: **a type of component value***:

> The syntax of a pseudo-class consists of a ":" (U+003A COLON) followed by the name of the pseudo-class as an [`<ident>`](https://drafts.csswg.org/css-values-4/#typedef-ident), or in the case of a functional pseudo-class, by a [functional notation](https://drafts.csswg.org/css-values-4/#functional-notation).

The [type definition of `<pseudo-class-selector>`](https://drafts.csswg.org/selectors/#typedef-pseudo-class-selector) in `17. Grammar` would be anchored to this prose in [3.5 Pseudo-classes](https://drafts.csswg.org/selectors/#pseudo-classes).

**Comment by svgeesus** [comment]
@cdoublev would you be willing to make a PR for this?

**Comment by cdoublev** [comment]
TLDR: you can probably close this issue right now, as I think you are primarily interested in moving Selectors to CR. Perhaps @tabatkins (who added the "Needs eedit" tag) want to confirm.

5c2f26b added how to [define a function using the CSS value definition syntax](https://drafts.csswg.org/css-values-4/#component-functions):

  > The syntax of a functional notation is defined as a sequence of:
  >
  > 1. The function’s name written as an identifier followed by an open parenthesis (such as `example()`, or **the `<function-token>` production to indicate a function with an arbitrary name**).
  > 2. The function’s arguments, if any, expressed using the value definition syntax.
  > 3. A literal closing parenthesis.

The CSS value definition syntax is explicitly defined as representing **component values**, which excludes `<function-token>`. So that seems baffling, or even inappropriate, for an implementer with little experience in CSS parsing/specs, who is tring to match a list of component values against such definition.

But this may be preferable to a definition in prose, for these functions whose name accepts a value range. The CSS value definition syntax does not allow representing them with eg. `<dashed-ident>(<any-value>)` (#10558). `<function>`, defined as having any value for its name and `<any-value>` for its value, is not much a better solution imo.

**Comment by tabatkins** [comment]
Yeah, I fixed up the VDS to allow this usage. The function token is technically gone by the time we validate grammars (replaced by a function *object* which contains its argument), but the VDS annotates a function *with* the function-token and closing parenthesis.

So this is indeed fixed now.
